torch/ao/quantization/dynamic_tracing/quantization_state.py

```diff
@@ -0,0 +1,797 @@
+            new_outputs = []
+            for output in outputs:
+                if isinstance(output, torch.Tensor):
+                    new_outputs.append(self._first_call_assign_qtensor_infos_to_mod_outputs_tensor(
+                        output, qtensor_id))
+                else:
+                    new_outputs.append(output)
+            # hacky check for collections.namedtuple, TODO improve this
+            # https://stackoverflow.com/questions/2166818/how-to-check-if-an-object-is-an-instance-of-a-namedtuple
+            if hasattr(outputs, '_fields'):
+                outputs = outputs.__class__(*new_outputs)
+            else:
+                outputs = tuple(new_outputs)
+        else:
+            pass
+        return outputs
+
+    def _maybe_mod_outputs_dtype_transform(
+        self,
+        outputs: Any,
+    ) -> Any:
+        """
+        Takes `outputs` which are about to be returned from this module
+        to the caller. If this module has restrictions on the dtypes of
+        tensors it has to return, does the dtype conversion. Otherwise,
+        does nothing.
+        """
+        if isinstance(outputs, torch.Tensor):
+            qtensor_info = self.output_qtensor_infos[0]
+            if self.output_dtypes is not None:
+                assert qtensor_info is not None
+                # check the output dtype, and do the conversion if needed
+                output_dtype = self.output_dtypes[0]
+                if qtensor_info.inf_dtype != output_dtype:
+                    assert output_dtype is torch.float, \
+                        'non-float output dtypes not handled yet'
+                    outputs = outputs.dequantize()
+            else:
+                # if no output dtype was specified, do nothing
+                pass
+
+        return outputs
+
+    def _first_call_op_prepare_before_hook_create_subgraphs_tensor(
+        self,
+        op: Callable,
+        arg: Any,
+        arg_idx: int,
+        input_observed_arg_idxs: Optional[List[int]],
+        arg_tensor_infos: List[Optional[QTensorInfo]],
+        func_output_dtype_type: FuncOutputDTypeType,
+        qtensor_id: List[int],
+    ) -> None:
+        """
+        Runs the prepare hook during first_call for individual
+        tensors. If the input argument is a tensor, this function is
+        called directly. If the input argument is an iterable such
+        as a list or a tuple, this function is called on each element of
+        the iteratble.
+        """
+        # TODO(next): fix this for torch.cat
+        if not isinstance(arg, torch.Tensor):
+            arg_tensor_infos.append(None)
+            return
+
+        # If a tensor does not have an ID, add it. This allows
+        # us to track inputs shared by multiple quantizeable modules.
+        if not hasattr(arg, '_qtensor_info'):
+            arg._qtensor_info = QTensorInfo(qtensor_id[0], arg.dtype)  # type: ignore[attr-defined]
+            qtensor_id[0] += 1
```

Comment:
`tensor_id` is primarily used to track which op connects to which op, to reason about how individual ops connect to subgraphs.  This is essential to the whole scheme working.

`tensor_id` is also used to track observers.  This use of the ID is an implementation detail, and it could be changed to use operator index + arg/kwarg/output index if we want.